Configuration files created for babel, eslint and webpack and respect…

Piyush3079 opened the pull request:
New directory structure used:
`js/src`: contains the javascript modularised code.
`js/src/functions`: specific functions to be used in js files like `db_search.js` `db_structure.js` etc.
`js/src/variables`: specific methods to bring global scope variables defined by php in the scope of webpack generated bundle.
`js/src/utils`: utility functions like `PMA_tooltip` `PMA_sprintf` `PMA_ajaxShowMessage` etc 

Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Before submitting pull request, please check that every commit:

- [ ] Has proper Signed-Off-By
- [ ] Has commit message which describes it
- [ ] Is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any new functionality is covered by tests

Piyush3079 committed:
Configuration files created for babel, eslint and webpack and respective modules added in package manager
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

mauriciofauth commented on package.json:
```diff
@@ -5,8 +5,13 @@
   "license": "GPL-2.0",
+  "scripts": {
+    "dev:wds": "webpack-dev-server --progress",
+    "prod:build": "rimraf lib dist && babel js/es6/ -d lib --ignore .test.js && cross-env NODE_ENV=production webpack -p --progress"
```
I got this error after running this command.
```
yarn run v1.6.0
$ rimraf lib dist && babel js/es6/ -d lib --ignore .test.js && cross-env NODE_ENV=production webpack -p --progress
/bin/sh: 1: babel: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

devenbansod replied:
I could also reproduce this.

Piyush3079 replied:
You need to install babel-cli globally.
npm install -g npm-cli
it will remove the error

devenbansod replied:
I would suggest that you add `babel-cli` as a devDependency and the command would start to work automatically.
Expecting babel-cli to be installed globally just adds one more unnecessary step in the process.

Piyush3079 replied:
Ya thanks
I have done that.

Piyush3079 replied:
I have made this change on my local repository and I will commit it with some more changes I am working on related to configuration.

mauriciofauth replied:
This is creating a `lib` directory at the root of the project. Is this expected?

MauricioFauth reviewed (commented):
I got some warnings after `yarn install`.
```
yarn install v1.6.0
[1/4] Resolving packages...
warning webpack-cli > jscodeshift > babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read babeljs.io/env to update! 
warning webpack-cli > webpack-addons > jscodeshift > babel-preset-es2015@6.24.1: 🙌  Thanks for using Babel: we recommend using babel-preset-env now: please read babeljs.io/env to update! 
warning webpack-cli > jscodeshift > nomnom@1.8.1: Package no longer supported. Contact support@npmjs.com for more info.
warning webpack-cli > webpack-addons > jscodeshift > nomnom@1.8.1: Package no longer supported. Contact support@npmjs.com for more info.
[2/4] Fetching packages...
info fsevents@1.2.3: The platform "linux" is incompatible with this module.
info "fsevents@1.2.3" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
```

devenbansod commented on webpack.config.babel.js:
```diff
@@ -0,0 +1,58 @@
+*/
+export default {
+    entry: [
+        './js/es6/sql.js',
```
@Piyush3079 Do we plan to add a new directory `es6` inside `js` to hold the source?

Piyush3079 replied:
Right now I have not decided the exact directory structure of the updated code to be used, I just have copied the files in the es6 directory for testing purpose.
Once the scripts compile the code in development as well as production mode successfully, I will finalise the exact directory structure.

devenbansod commented on webpack.config.babel.js:
```diff
@@ -0,0 +1,58 @@
+    resolve: {
+        extensions: ['.js'],
+    },
+    devServer: {
```
Does the webpack server use `node` server inside it? If so, would it be able to serve `phpMyAdmin` as it is (since it's a PHP application)? 
If no, what's the use of dev-server here?

Piyush3079 replied:
yes it uses the `node` server for serving the compiled files on the go.
We don't need to manually compile the code again and again after making small changes, the dev sever will compile the code on saving the file we are working on and will make the changes reflect on refreshing the browser..

shucon commented on .babelrc:
```diff
@@ -0,0 +1,3 @@
+{
+    "presets": ["env"]
+}
```
I think there should be a new line at the end of file.

Piyush3079 committed:
Test sql file added and rimraf package added for testing
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
More js files added as webpack configuratin and some options commented to reduce compilation time
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 commented on package-lock.json:
```diff
@@ -0,0 +1,6856 @@
+{
+  "name": "phpmyadmin",
```
`yarn.lock` ensures consistent install of dependencies. So having it on source control ensures the installation of exact dependencies tree across all machines.
https://yarnpkg.com/lang/en/docs/yarn-lock/#toc-check-into-source-control

devenbansod replied:
Apologies for not being clear in my comment. 

I understand that and do completely agree. But then since we are using `yarn` and not `npm`, why is `package-lock.json` also added?

Piyush3079 replied:
Oh, that might be uploaded by mistake, I will remove it, thanks for pointing out.

Piyush3079 committed:
package-lock.json removed and file name of menu-resizer changed for webpack
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
new webpack configuration for exposing global variables as libraries
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Header and Script class edited for new directory structure and global variables are edited
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
exposing some global methods to take parameters defined in js/src/messages.php
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
messages.php edited for using globally exposed methods defined in js/src/global/pma_messages.js
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Some changes in the configuration to revamp file by file not whole directory, environment variable introduced
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
db_search.js added with some es6 features and trailing white spaces removed
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
documentation updated for environment variables
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

devenbansod commented on doc/config.rst:
```diff
@@ -3391,6 +3391,21 @@ Developer
+    :default: 3307
+
+    Enable you to start webpack development server for serving js files
+    when woking in the development environment
```
/minor typo

Piyush3079 committed:
Some functions moved from server_privilages.js to functions.js and typo error removed
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
New directory Stucture having modules imported and exported.
This commit needs proper discussion as many of the upcoming changes are based on this commit.
These files contains the working code only.
Right now the code has the functionality to work with both old js files and new js files.
Commeting style is yet to decide
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

mauriciofauth commented on config.sample.inc.php:
```diff
@@ -16,6 +16,20 @@
+ * are in development environment
+ */
+$cfg['webpack_port'] = 3307;
+
```
I think these directives should be in `libraries/config.default.php` file.

Piyush3079 replied:
But if someone want to change these configurations then he has to edit the `config.default.php` and we need to include this in the documentation that if you are a developer then you need to first change the environment in default config file.
How can we deal with this??

devenbansod replied:
Hi @Piyush3079 the values defined in `config.inc.php` are meant to override the ones defined in `libraries/config.default.php`.

Ref: https://docs.phpmyadmin.net/en/latest/config.html#configuration

mauriciofauth commented on doc/config.rst:
```diff
@@ -3391,6 +3391,21 @@ Developer
+    :type: string
+    :default: `'production'`
+
+    Enable you to set the working environment for loading js files
```
Instead of describing it as something related to JS files, you could describe it as something more generic.
What do you think?
You could also comment on the possible values (`'production'` and `'development'`).

Piyush3079 replied:
We can remove the js part and have options like enable js debugging, start webpack dev server, sql logging and other debugging under `development` environment and if selected environment is `development` then he can enable other logging and debugging features.

Piyush3079 commented:
Ok, got it, I will make required changes in `config.default.php`.

Piyush3079 committed:
server_privilages and server_databases modularized
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
AJAX handler modified and initial loading of the first page handled
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 commented on js/src/index.js:
```diff
@@ -0,0 +1,112 @@
+ * on pages and submissions of forms
+ */
+$(document).on('click', 'a', AJAX.requestHandler);
+$(document).on('submit', 'form', AJAX.requestHandler);
```
@devenbansod @mauriciofauth 
I have used this code from ajax.js (the old one), so that all the ajax requests will be handled by new ajax file but this is sending the requests of other `<a>` clicks also to this handler instead of sending them to their respective event handler which were created due the loading of those files.

devenbansod replied:
I believe the function AJAX.requestHandler in ajax.js is always called even currently, but [this code](https://github.com/phpmyadmin/phpmyadmin/blob/master/js/ajax.js#L218-L235) might be getting the event handling out of that specific function and move to the individual event handlers in different files.

Piyush3079 replied:
Actually I am working on error_report.js and there in [this line](https://github.com/phpmyadmin/phpmyadmin/blob/86fcc46c32ac880dad7b83f39007d4c1c97cc845/js/error_report.js#L149), it is creating an event to redirect to settings page but initially the request was sent to ajax handler and that too up to ajax response handler and then on again clicking, is redirecting to settings page.
The conditions were not able to handle this click event.
Well I am working on it and try to fix it.

Piyush3079 committed:
PMA console broken up into classes and imported.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Error report for js errors tested for ajax callback function.
TODO: To find a way to wrap the exported function containing PMA_ into the error handlerr to generate backtace.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Revamped server_status_advisor, server_status_variables, sever_status_processes and codemirror package added in server_privileges.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Navigation revamped, navigation functions separated from main file. More file names added in ajax and index.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
File names changed for revamped javascript files in php files.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 commented:
Actually right now I have not changed the names of functions and objects but as soon I finish the remaining files being revamped, I will start worlking on decreasing the issue count on scruitinizer as most of the issues are related to camel casing and apart from that this `PMA_` is being used in `ErrorReport.js` to wrap the functions starting with `PMA_` and invoke them by `try catch` wrapper. This is helping the `TraceKit` to geenrate the error report (as far as I have understood the code).
Please do correct me if I am wrong regarding the wrapping explaination.

Piyush3079 committed:
Bugs related to server status processes removed
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Index files edited and more files whitelisted in ajax.js and files.js.
Navigation files edited for validation of navigation settings,
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Server related remaining files modularised like replication, plugins, queries, processes.
Sever monitor is also modularised but is not working as jQplot is giving error with webpack. So I am still working on this.
Another issue with these files is that table sorter is also not woking. I am working on this too.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Config.js is modularised.
Signed-off-by: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Test files changed as per the new chages in FormDisplayTemplate.
Js code is stuctured properly with comments.
Signed-Off-By Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Some more javascript files structured properly.
Shortcut handler and page settings js files added in modular code.
Signed-Off-By: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
Weekly progress.
1). Microhistory added in classes for modular code
2). Code is structured and camel casing of varables is done for removing linter warnings.
3). JQuery plugins are added in some files.
4). jQplot is working for server_status_queries and server_status_monitor.
5). Work started for Dabatase, SQl and Table files.
Some changes are stll left for server_status_monitor and sql.js
Signed-Off-By: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 committed:
AJAX handler, Header.php and Scripts.php modified for compatibility with modularized and non-modularized code.
Signed-Off-By: Piyush Vijay <piyushvijay.1997@gmail.com>

Piyush3079 commented:
This pr contains lots of commented code and also this pr is very large to be reviewed in one go so I am closing this pr and I have made several prs from this pr which contains changes that can be reviewed.
The prs created from this branch in continuation are:
Mod_Js_Config
Mod_Js_Server_Privileges
Mod_Js_Console
Mod_Js_Server_Status
Mod_Js_Navigation
Mod_Js_Import_Export
